File: core/src/com/ultreon/craft/resources/Identifier.java
package com.ultreon.craft.resources;

import com.ultreon.craft.UltreonCraft;

import java.util.Objects;
import java.util.regex.Pattern;

public final class Identifier {
    private static final String DEFAULT_NAMESPACE = UltreonCraft.NAMESPACE;
    private static final Pattern namespaceRegex = Pattern.compile("[a-z0-9_\\-]*");
    private static final Pattern pathRegex = Pattern.compile("[a-z0-9_\\-/.]*");
    private final String namespace;
    private final String path;

    public Identifier(String text) {
        if (text.contains(":")) {
            var split = text.split(":", 2);
            namespace = testNamespace(split[0]);
            path = testPath(split[1]);
        } else {
            namespace = DEFAULT_NAMESPACE;
            path = testPath(text);
        }
    }

    public Identifier(String namespace, String path) {
        this.namespace = namespace;
        this.path = path;
    }

    private String testNamespace(String namespace) {
        if (namespaceRegex.matcher(namespace).find()) {
            return namespace;
        }

        throw new IllegalArgumentException("Invalid namespace: $namespace");
    }

    private String testPath(String path) {
        if (namespaceRegex.matcher(path).find()) {
            return path;
        }

        throw new IllegalArgumentException("Invalid path: $path");
    }

    public String namespace() {
        return namespace;
    }

    public String path() {
        return path;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) return true;
        if (obj == null || obj.getClass() != this.getClass()) return false;
        var that = (Identifier) obj;
        return Objects.equals(this.namespace, that.namespace) &&
                Objects.equals(this.path, that.path);
    }

    @Override
    public int hashCode() {
        return Objects.hash(namespace, path);
    }

    @Override
    public String toString() {
        return namespace + "@" + path;
    }
}

File: src/jvmMain/kotlin/com/ultreon/craft/resources/Identifier.kt
package com.ultreon.craft.resources

data class Identifier(val namespace: String, val path: String) {
    override fun toString(): String {
        return "$namespace:$path"
    }

    companion object {
        private val namespaceRegex = Regex("[a-z0-9_\\-]*")
        private val pathRegex = Regex("[a-z0-9_\\-/.]*")

        private fun testNamespace(namespace: String): String {
            if (namespaceRegex.matches(namespace)) {
                return namespace
            }

            throw IllegalArgumentException("Invalid namespace: $namespace")
        }

        private fun testPath(path: String): String {
            if (namespaceRegex.matches(path)) {
                return path
            }

            throw IllegalArgumentException("Invalid path: $path")
        }

        operator fun invoke(text: String): Identifier {
            if (":" in text) {
                val split = text.split(":", limit = 2)
                return Identifier(testNamespace(split[0]), testPath(split[1]))
            }
            return Identifier(defaultNamespace, testPath(text))
        }
    }
}

fun id(namespace: String, path: String) = Identifier(namespace, path)
fun id(text: String) = Identifier(text)
fun craftId(text: String) = Identifier(defaultNamespace, text)
fun idOrNull(text: String): Identifier? = try {
    Identifier(text)
} catch (e: IllegalArgumentException) {
    null
}

const val defaultNamespace: String = "craft"
